Refines ScopeValidator Diagnostics to use more specific Diagnostic tags

markwpearce opened the pull request:
Fixes an issue where more diagnostics would be cleared on file changes than should have been.

Fixes #1209

TwitchBronBron commented on src/bscPlugin/validation/ScopeValidator.ts:
```diff
@@ -60,9 +60,10 @@ export class ScopeValidator {
             return;
         }
-        this.event.program.diagnostics.clearByFilter({ scope: this.event.scope, tag: ScopeValidatorDiagnosticTag });
+        this.event.program.diagnostics.clearByFilter({ scope: this.event.scope, segment: null, tag: ScopeValidatorDiagnosticTag });
```
Setting these to null is a bit unintuitive, especially combined with the `!== undefined` above. Is this intentional? Throughout the codebase, we've tried to basically treat `null` and `undefined` as the same thing. 

What does setting it to `null` accomplish? Does it say "don't match on segements"? Or "mach all segments"? I'm wondering if there's a more clear way to handle this, since plugin authors will also be calling these functions.

TwitchBronBron commented on src/DiagnosticManager.ts:
```diff
@@ -197,14 +197,15 @@ export class DiagnosticManager {
-            scope: !!filter.scope,
-            file: !!filter.file,
-            segment: !!filter.segment
+            tag: !!(filter.tag),
+            scope: !!(filter.scope),
+            file: !!(filter.file),
+            segment: !!(filter.segment)
```
Are the parentheses needed now?

markwpearce replied:
nope .. i'll remove them

markwpearce commented:
Changed tack here... Instead of being laisez-faire about the diagnostics clearing, instead i've changed it to be very specific to the what is cleared.. so each validation function can decide best on how to clear existing diagnostics
